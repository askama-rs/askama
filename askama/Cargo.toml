[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
