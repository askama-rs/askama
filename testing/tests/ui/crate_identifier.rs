use askama::Template;

#[derive(Template)]
#[template(ext = "html", source = "{{ crate }}")]
struct Crate;

#[derive(Template)]
#[template(ext = "html", source = "{% if crate == 12 %}{% endif %}")]
struct Crate2;

#[derive(Template)]
#[template(ext = "html", source = "{% match crate %}{% endmatch %}")]
struct Crate3;

#[derive(Template)]
#[template(ext = "html", source = "{% let crate %}")]
struct Crate4;

#[derive(Template)]
#[template(ext = "html", source = "{% let crate = 12 %}")]
struct Crate5;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.a.crate }}")]
struct Crate6 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{{ super }}")]
struct Super;

#[derive(Template)]
#[template(ext = "html", source = "{% if super == 12 %}{% endif %}")]
struct Super2;

#[derive(Template)]
#[template(ext = "html", source = "{% match super %}{% endmatch %}")]
struct Super3;

#[derive(Template)]
#[template(ext = "html", source = "{% let super %}")]
struct Super4;

#[derive(Template)]
#[template(ext = "html", source = "{% let super = 12 %}")]
struct Super5;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.a.super }}")]
struct Super6 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{{ Self }}")]
struct Self1;

#[derive(Template)]
#[template(ext = "html", source = "{% if Self == 12 %}{% endif %}")]
struct Self2;

#[derive(Template)]
#[template(ext = "html", source = "{% match Self %}{% endmatch %}")]
struct Self3;

#[derive(Template)]
#[template(ext = "html", source = "{% let Self %}")]
struct Self4;

#[derive(Template)]
#[template(ext = "html", source = "{% let Self = 12 %}")]
struct Self5;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.a.Self }}")]
struct Self6 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% let self %}")]
struct SmallSelf;

#[derive(Template)]
#[template(ext = "html", source = "{% let self = 12 %}")]
struct SmallSelf2;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.a.self }}")]
struct SmallSelf3 {
    a: u8,
}

// Regression test for <https://github.com/askama-rs/askama/issues/449>.
#[derive(Template)]
#[template(
    ext = "",
    source = "{{\u{c}KK3e331<c7}}61/63m3333u7<c0.}}\u{6}\0\u{c}\u{c}{{c/crate<338}}6unsafe/63a3ae\u{c}\u{c}\u{c}%et"
)]
struct Regression {}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when a::b::super %}{% endmatch %}")]
struct PathElemSuper {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when self::a::b::super %}{% endmatch %}")]
struct PathElemSuper2 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when a::b::self %}{% endmatch %}")]
struct PathElemSelf {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when self::a::b::self %}{% endmatch %}")]
struct PathElemSelf2 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when a::b::crate %}{% endmatch %}")]
struct PathElemCrate {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when self::a::b::crate %}{% endmatch %}")]
struct PathElemCrate2 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when a::b::Self %}{% endmatch %}")]
struct PathElemSelfType {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{% match a %}{% when self::a::b::Self %}{% endmatch %}")]
struct PathElemSelfType2 {
    a: u8,
}

#[derive(Template)]
#[template(ext = "html", source = "{{ Self.4 }}")]
struct InvalidRawIdentifierSelfTyInAttr1;

#[derive(Template)]
#[template(ext = "html", source = "{{ super.4 }}")]
struct InvalidRawIdentifierSuperInAttr1;

#[derive(Template)]
#[template(ext = "html", source = "{{ crate.4 }}")]
struct InvalidRawIdentifierCrateInAttr1;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.Self.4 }}")]
struct InvalidRawIdentifierSelfTyInAttr2;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.super.4 }}")]
struct InvalidRawIdentifierSuperInAttr2;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.crate.4 }}")]
struct InvalidRawIdentifierCrateInAttr2;

#[derive(Template)]
#[template(ext = "html", source = "{{ self.self.4 }}")]
struct InvalidRawIdentifierSelfInAttr;

fn main() {}
