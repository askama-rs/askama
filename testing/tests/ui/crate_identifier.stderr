error: `crate` cannot be used as an identifier
 --> <source attribute>:1:3
       "crate }}"
 --> tests/ui/crate_identifier.rs:4:35
  |
4 | #[template(ext = "html", source = "{{ crate }}")]
  |                                   ^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:6
       "crate == 12 %}{% endif %}"
 --> tests/ui/crate_identifier.rs:8:35
  |
8 | #[template(ext = "html", source = "{% if crate == 12 %}{% endif %}")]
  |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:9
       "crate %}{% endmatch %}"
  --> tests/ui/crate_identifier.rs:12:35
   |
12 | #[template(ext = "html", source = "{% match crate %}{% endmatch %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: cannot use `crate` as a name: it is a rust keyword
 --> <source attribute>:1:7
       "crate %}"
  --> tests/ui/crate_identifier.rs:16:35
   |
16 | #[template(ext = "html", source = "{% let crate %}")]
   |                                   ^^^^^^^^^^^^^^^^^

error: cannot use `crate` as a name: it is a rust keyword
 --> <source attribute>:1:7
       "crate = 12 %}"
  --> tests/ui/crate_identifier.rs:20:35
   |
20 | #[template(ext = "html", source = "{% let crate = 12 %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:10
       "crate }}"
  --> tests/ui/crate_identifier.rs:24:35
   |
24 | #[template(ext = "html", source = "{{ self.a.crate }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:3
       "super }}"
  --> tests/ui/crate_identifier.rs:30:35
   |
30 | #[template(ext = "html", source = "{{ super }}")]
   |                                   ^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:6
       "super == 12 %}{% endif %}"
  --> tests/ui/crate_identifier.rs:34:35
   |
34 | #[template(ext = "html", source = "{% if super == 12 %}{% endif %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:9
       "super %}{% endmatch %}"
  --> tests/ui/crate_identifier.rs:38:35
   |
38 | #[template(ext = "html", source = "{% match super %}{% endmatch %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: cannot use `super` as a name: it is a rust keyword
 --> <source attribute>:1:7
       "super %}"
  --> tests/ui/crate_identifier.rs:42:35
   |
42 | #[template(ext = "html", source = "{% let super %}")]
   |                                   ^^^^^^^^^^^^^^^^^

error: cannot use `super` as a name: it is a rust keyword
 --> <source attribute>:1:7
       "super = 12 %}"
  --> tests/ui/crate_identifier.rs:46:35
   |
46 | #[template(ext = "html", source = "{% let super = 12 %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:10
       "super }}"
  --> tests/ui/crate_identifier.rs:50:35
   |
50 | #[template(ext = "html", source = "{{ self.a.super }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:3
       "Self }}"
  --> tests/ui/crate_identifier.rs:56:35
   |
56 | #[template(ext = "html", source = "{{ Self }}")]
   |                                   ^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:6
       "Self == 12 %}{% endif %}"
  --> tests/ui/crate_identifier.rs:60:35
   |
60 | #[template(ext = "html", source = "{% if Self == 12 %}{% endif %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:9
       "Self %}{% endmatch %}"
  --> tests/ui/crate_identifier.rs:64:35
   |
64 | #[template(ext = "html", source = "{% match Self %}{% endmatch %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:7
       "Self %}"
  --> tests/ui/crate_identifier.rs:68:35
   |
68 | #[template(ext = "html", source = "{% let Self %}")]
   |                                   ^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:7
       "Self = 12 %}"
  --> tests/ui/crate_identifier.rs:72:35
   |
72 | #[template(ext = "html", source = "{% let Self = 12 %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:10
       "Self }}"
  --> tests/ui/crate_identifier.rs:76:35
   |
76 | #[template(ext = "html", source = "{{ self.a.Self }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^

error: cannot use `self` as a name: it is a rust keyword
 --> <source attribute>:1:7
       "self %}"
  --> tests/ui/crate_identifier.rs:82:35
   |
82 | #[template(ext = "html", source = "{% let self %}")]
   |                                   ^^^^^^^^^^^^^^^^

error: cannot use `self` as a name: it is a rust keyword
 --> <source attribute>:1:7
       "self = 12 %}"
  --> tests/ui/crate_identifier.rs:86:35
   |
86 | #[template(ext = "html", source = "{% let self = 12 %}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^

error: `self` cannot be used as an identifier
 --> <source attribute>:1:10
       "self }}"
  --> tests/ui/crate_identifier.rs:90:35
   |
90 | #[template(ext = "html", source = "{{ self.a.self }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:41
       "crate<338}}6unsafe/63a3ae\u{c}\u{c}\u{c}%et"
  --> tests/ui/crate_identifier.rs:99:14
   |
99 |     source = "{{\u{c}KK3e331<c7}}61/63m3333u7<c0.}}\u{6}\0\u{c}\u{c}{{c/crate<338}}6unsafe/63a3ae\u{c}\u{c}\u{c}%et"
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:27
       "super %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:104:35
    |
104 | #[template(ext = "html", source = "{% match a %}{% when a::b::super %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:33
       "super %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:110:35
    |
110 | #[template(ext = "html", source = "{% match a %}{% when self::a::b::super %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `self` cannot be used as an identifier
 --> <source attribute>:1:27
       "self %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:116:35
    |
116 | #[template(ext = "html", source = "{% match a %}{% when a::b::self %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `self` cannot be used as an identifier
 --> <source attribute>:1:33
       "self %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:122:35
    |
122 | #[template(ext = "html", source = "{% match a %}{% when self::a::b::self %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:27
       "crate %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:128:35
    |
128 | #[template(ext = "html", source = "{% match a %}{% when a::b::crate %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:33
       "crate %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:134:35
    |
134 | #[template(ext = "html", source = "{% match a %}{% when self::a::b::crate %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:27
       "Self %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:140:35
    |
140 | #[template(ext = "html", source = "{% match a %}{% when a::b::Self %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:33
       "Self %}{% endmatch %}"
   --> tests/ui/crate_identifier.rs:146:35
    |
146 | #[template(ext = "html", source = "{% match a %}{% when self::a::b::Self %}{% endmatch %}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:3
       "Self.4 }}"
   --> tests/ui/crate_identifier.rs:152:35
    |
152 | #[template(ext = "html", source = "{{ Self.4 }}")]
    |                                   ^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:3
       "super.4 }}"
   --> tests/ui/crate_identifier.rs:156:35
    |
156 | #[template(ext = "html", source = "{{ super.4 }}")]
    |                                   ^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:3
       "crate.4 }}"
   --> tests/ui/crate_identifier.rs:160:35
    |
160 | #[template(ext = "html", source = "{{ crate.4 }}")]
    |                                   ^^^^^^^^^^^^^^^

error: `Self` cannot be used as an identifier
 --> <source attribute>:1:8
       "Self.4 }}"
   --> tests/ui/crate_identifier.rs:164:35
    |
164 | #[template(ext = "html", source = "{{ self.Self.4 }}")]
    |                                   ^^^^^^^^^^^^^^^^^^^

error: `super` cannot be used as an identifier
 --> <source attribute>:1:8
       "super.4 }}"
   --> tests/ui/crate_identifier.rs:168:35
    |
168 | #[template(ext = "html", source = "{{ self.super.4 }}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^

error: `crate` cannot be used as an identifier
 --> <source attribute>:1:8
       "crate.4 }}"
   --> tests/ui/crate_identifier.rs:172:35
    |
172 | #[template(ext = "html", source = "{{ self.crate.4 }}")]
    |                                   ^^^^^^^^^^^^^^^^^^^^

error: `self` cannot be used as an identifier
 --> <source attribute>:1:8
       "self.4 }}"
   --> tests/ui/crate_identifier.rs:176:35
    |
176 | #[template(ext = "html", source = "{{ self.self.4 }}")]
    |                                   ^^^^^^^^^^^^^^^^^^^
